Bookmarks - Multiselect support

ashiagr opened the pull request:
## Description

This adds the multi-select ability (Io7PCz1H1hmOgEE9eljoYW-fi-2033_32707) to the bookmark list for Android. 

These commits refactor `MultiSelectHelper` and `MultiSelectAction` to support bookmarks:
0375332dd380544d3ddf20f63cff75aef2f4735b...b8e2894c8108a66405758bf4d5e12f78ff353188

Some of the functions needed for both bookmarks and episodes are moved to a parent class and code movement has increased PR size. It might be good to test them as a whole but I can split the PR if there's any issue reviewing it. 

## Testing Instructions

#### Test.1 Multi-Select Mode 

1. Edit the `BookmarkManagerImpl` class to return a flow of (multiple) bookmarks from `findEpisodeBookmarksFlow(...)`
2. Install the debug build and login with a `Plus` account
3. Open the Full Screen Player
4. Swipe over to Bookmarks
5. Long press on a row
6. ✅ Verify you enter multi selection mode and the action bar appears
7. Tap on a row, ✅ verify the item is now selected
8. ✅ Verify the action bar selection count updated
9. ✅ Verify the edit option is visible
10. Tap the row again to deselect it, ✅ Verify the edit option is hidden
11. Tap Select All from more options in the multi-select toolbar
12. ✅ Verify all the items are selected
13. Tap and hold on a row again
14. ✅ Verify you see the 'Select All Above' and 'Select All Below' options
15. Tap each option and ✅ verify the correct items are selected

#### Test.2  ViewPager swipe action and Player close
1. With the multi-select toolbar visible, swipe to the right to go the previous page
2. ✅ Notice that `ViewPager` swipe action is disabled
4. With the multi-select toolbar visible, apply the back gesture
5. ✅ Notice that multi-select mode ends
6. Swipe again to the right to go the previous page
7. ✅ Notice that `ViewPager` swipe action is enabled and can be swiped to the previous page
8. Go back to the `Bookmarks` tab and re-enable multi-select mode
9. Close the full screen player and reopen it
10. ✅ Notice that multi-select toolbar is not visible

## Screenshots or Screencast 


https://github.com/Automattic/pocket-casts-android/assets/1405144/cc55b331-cf41-4efe-9801-daab30bad3f2


## Checklist
- [ ] If this is a user-facing change, I have added an entry in CHANGELOG.md
- [ ] I have considered whether it makes sense to add tests for my changes
- [x] All strings that need to be localized are in `modules/services/localization/src/main/res/values/strings.xml`
- [x] Any jetpack compose components I added or changed are covered by compose previews
 
#### I have tested any UI changes...
<!-- If this PR does not contain UI changes, ignore these items -->
- [x] with different themes
- [x] with a landscape orientation
- [x] with the device set to have a large display and font size
- [ ] for accessibility with TalkBack

ashiagr commented on modules/services/views/src/main/java/au/com/shiftyjelly/pocketcasts/views/multiselect/MultiSelectHelper.kt:
```diff
@@ -8,181 +8,100 @@ import androidx.fragment.app.FragmentManager
-        get() = Dispatchers.Default
+    open lateinit var listener: Listener<T>
 
     private val _isMultiSelectingLive = MutableLiveData<Boolean>().apply { value = false }
     val isMultiSelectingLive: LiveData<Boolean> = _isMultiSelectingLive
 
+    protected val selectedList: MutableList<T> = mutableListOf()
+    protected val _selectedListLive = MutableLiveData<List<T>>().apply { value = listOf() }
+    val selectedListLive: LiveData<List<T>> = _selectedListLive
+    val selectedCount: LiveData<Int> = _selectedListLive.map { it.size }
```
I haven't converted existing`LiveData` to `Flow` as it would further increase PR size. I've converted them to `Flow` in the `BookmarkViewModel`, mocking `LiveData` from `MultiSelectHelper` in the test class is not working well so I've commented out tests in [be86cb5](https://github.com/Automattic/pocket-casts-android/pull/1173/commits/be86cb507eb048a6008babf4fa9cbfd9ccf97bd9). 

I'll try to fix it in a follow up PR.

ashiagr commented on modules/services/views/src/main/java/au/com/shiftyjelly/pocketcasts/views/multiselect/MultiSelectHelper.kt:
```diff
@@ -8,181 +8,100 @@ import androidx.fragment.app.FragmentManager
-) :
-    CoroutineScope {
-    interface Listener {
+abstract class MultiSelectHelper<T> : CoroutineScope {
```
Parent `MultiSelectHelper<T>` uses generics. H/t @geekygecko for the suggestion.

ashiagr commented on modules/services/views/src/main/java/au/com/shiftyjelly/pocketcasts/views/multiselect/MultiSelectAction.kt:
```diff
@@ -108,59 +26,185 @@ sealed class MultiSelectAction(
+        )
+    }
+
+    sealed class BookmarkAction(
```
Grouped episode and bookmark actions inside `BaseEpisodeAction` and `BookmarkAction` respectively.

ashiagr commented on modules/features/player/src/main/res/layout/fragment_player_container.xml:
```diff
@@ -9,74 +9,87 @@
+
+            </LinearLayout>
+
+            <au.com.shiftyjelly.pocketcasts.views.multiselect.MultiSelectToolbar
+                android:id="@+id/multiSelectToolbar"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:minHeight="?android:attr/actionBarSize"
+                android:theme="@style/ThemeDark"
+                android:visibility="gone"/>
```
Reuses existing `MultiSelectToolbar` `Android` view. Thankfully, it interoperates well with the bookmarks `Compose` view. 😅

geekygecko commented on modules/services/views/src/main/java/au/com/shiftyjelly/pocketcasts/views/multiselect/MultiSelectBookmarksHelper.kt:
```diff
@@ -0,0 +1,60 @@
+                true
+            }
+
+            R.id.menu_unselect_all -> {
```
I see there is an unselect all but if all options are selected this action isn't shown.

ashiagr replied:
Removed in c109ee0 for now. Might need it again for the deselect changes in https://github.com/Automattic/pocket-casts-android/pull/1172

geekygecko commented on modules/services/views/src/main/java/au/com/shiftyjelly/pocketcasts/views/multiselect/MultiSelectToolbar.kt:
```diff
@@ -30,42 +30,54 @@ class MultiSelectToolbar @JvmOverloads constructor(
-                it.subList(0, MAX_ICONS).forEachIndexed { _, action ->
+                val maxIcons = when (multiSelectHelper) {
+                    is MultiSelectBookmarksHelper -> MAX_ICONS_BOOKMARKS
+                    is MultiSelectEpisodesHelper -> MAX_ICONS
```
Just an idea so feel free to ignore it. Would it be better to keep the bookmarks and episodes out of this class. Instead the multiSelectHelper could have an abstract method.
```
val maxIcons = multiSelectHelper.maxIcons
...
abstract val maxIcons: Int
```

ashiagr replied:
Nice idea! I applied it in 5a98c7c.

geekygecko commented on modules/services/views/src/main/java/au/com/shiftyjelly/pocketcasts/views/multiselect/MultiSelectToolbar.kt:
```diff
@@ -30,42 +30,54 @@ class MultiSelectToolbar @JvmOverloads constructor(
+                        val overflow = menu.add(Menu.NONE, R.id.menu_overflow, 0, context.getString(LR.string.more_options))
+                        overflow.setIcon(IR.drawable.ic_more_vert_black_24dp)
+                        overflow.setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS)
+                    }
```
Why is there different logic for bookmarks and episodes? I'm not quite following this part.

ashiagr replied:
It's a single menu item, and doesn't need any edit action, that's why.

geekygecko commented on modules/services/views/src/main/java/au/com/shiftyjelly/pocketcasts/views/multiselect/MultiSelectAction.kt:
```diff
@@ -108,59 +26,185 @@ sealed class MultiSelectAction(
+        @DrawableRes override val iconRes: Int,
+        override val analyticsValue: String,
+        override val isVisible: Boolean = true,
+    ) : MultiSelectAction(
```
It feels a bit strange that the episodes and bookmarks actions are in the same file. I'm not sure how to fix it, maybe a MultiSelectBookmarkActions and MultiSelectEpisodeActions class. 🤔

ashiagr replied:
Separated classes in 7dde7d0

geekygecko reviewed (approved):
This worked well for me. You have done a great job of reusing existing logic. I only had a couple of comments, feel free to ignore them.

ashiagr commented:
Thanks for the review comments @geekygecko! I've addressed/ responded to those.

Just FYI, I was facing an issue with merging `main` with this branch so I rebased my commits in c109ee0.
